### ISSUE TITLE
Faster BroadPhase

### ISSUE THREAD

**Issue by erincatto** [erincatto/box2d#485]
Instead of using sorting the pair buffer and skipping duplicates, just use the move flag.

**Comment by erincatto** [comment]
I looked at this more closely. The performance trade-off is between sorting the pair buffer and clearing the move flags. This trade-off is not clear to me since clearing flags can lead to cache misses.

It might be better to use a time stamp that gets incremented once per time step in b2World. Each proxy would also have a time stamp. Any proxies that move would sync their time stamp to the latest world time stamp. Then we could easily test if a proxy moved in the current time step.

So this avoids the need to clear a move flag.

Downside: in the case of time stamp overflow, all proxies would need to have their time stamps reset.

**Comment by louis-langholtz** [comment]
I've tried putting a moved flag into the body instances. This may cut down on some calls to `b2Body::SynchronizeFixtures()` and it appears that my clear flag calls can be done along side other calls that similarly need to update the body anyway.

I haven't found clear indication of speedup though.

Regardless, I do prefer using this flag as the signal for synchronizing with the broad phase over whether it was islanded and not static in the `Solve` and `SolveTOI` contexts.

Can't say I see value in putting a moved flag or time stamp field into the dynamic tree nodes. Won't the `b2ContactManager::FindNewContacts()` call path still need to do pairings and sortings and make them unique (in `b2BroadPhase::UpdatePairs`)?

Avoiding the sort and unique actions does seem like it'd be a win for performance but I'm not seeing how to remove those.

I have also changed `b2ContactManager::AddPair` code to check which body (`bodyA` or `bodyB`) has the least contacts and to use that body to check if the contact already exists. I believe I did see some speedup in some circumstances with that change. I did other things too though which makes it less clear to me how much performance improvement picking the less contacted body actually made. Like I had also changed the body contacts list to be an array of elements where each element was a structure containing the sorted proxy ID pair along with the contact pointer so the searches didn't necessarily have to read the contact data which could be colder (cache-wise).
